clang/lib/Driver/Driver.cpp

```diff
@@ -1344,206 +1285,189 @@ void Driver::CreateOffloadingDeviceToolChains(Compilation &C,
-              FoundNormalizedTriples.try_emplace(NormalizedName, Triple);
+  bool IsSYCL = C.getInputArgs().hasFlag(options::OPT_fsycl,
+                                         options::OPT_fno_sycl, false) ||
+                C.getInputArgs().hasArgNoClaim(options::OPT_fsycl_device_only,
+                                               options::OPT_fsyclbin_EQ);
+  bool IsOpenMPOffloading =
+      UseLLVMOffload ||
+      (C.getInputArgs().hasFlag(options::OPT_fopenmp, options::OPT_fopenmp_EQ,
+                                options::OPT_fno_openmp, false) &&
+       (C.getInputArgs().hasArg(options::OPT_offload_targets_EQ) ||
+        (C.getInputArgs().hasArg(options::OPT_offload_arch_EQ) &&
+         !(IsCuda || IsHIP))));
+
+  llvm::SmallSet<Action::OffloadKind, 4> Kinds;
+  const std::pair<bool, Action::OffloadKind> ActiveKinds[] = {
+      {IsCuda, Action::OFK_Cuda},
+      {IsHIP, Action::OFK_HIP},
+      {IsOpenMPOffloading, Action::OFK_OpenMP},
+      {IsSYCL, Action::OFK_SYCL}};
+  for (const auto &[Active, Kind] : ActiveKinds)
+    if (Active)
+      Kinds.insert(Kind);
+
+  // We currently don't support any kind of mixed offloading.
+  if (Kinds.size() > 1 && !IsSYCL) {
+    Diag(clang::diag::err_drv_mix_offload)
+        << Action::GetOffloadKindName(*Kinds.begin()).upper()
+        << Action::GetOffloadKindName(*(++Kinds.begin())).upper();
+    return;
+  }
 
-          if (!Inserted) {
-            Diag(clang::diag::warn_drv_sycl_offload_target_duplicate)
-                << Triple << TripleIt->second;
-            continue;
-          }
+  diagnoseSYCLOptions(C, IsSYCL);
+  // Initialize the compilation identifier used for unique CUDA / HIP names.
+  if (IsCuda || IsHIP)
+    CUIDOpts = CUIDOptions(C.getArgs(), *this);
 
-          // If the specified target is invalid, emit a diagnostic.
-          if (!isValidSYCLTriple(TT)) {
-            Diag(clang::diag::err_drv_invalid_sycl_target) << Triple;
-            continue;
-          }
+  // Get the list of requested offloading toolchains. If they were not
+  // explicitly specified we will infer them based on the offloading language
+  // and requested architectures.
+  std::multiset<llvm::StringRef> Triples;
+  if (C.getInputArgs().hasArg(options::OPT_offload_targets_EQ)) {
+    std::vector<std::string> ArgValues =
+        C.getInputArgs().getAllArgValues(options::OPT_offload_targets_EQ);
+    llvm::Triple TT;
+    for (llvm::StringRef Target : ArgValues) {
+      if(IsSYCL) {
+        if (Target.starts_with("intel_gpu_"))
+        Triples.insert(
+            C.getInputArgs().MakeArgString("spir64_gen-unknown-unknown"));
+        else if (Target.starts_with("nvidia_gpu_"))
+          Triples.insert(C.getInputArgs().MakeArgString("nvptx64-nvidia-cuda"));
+        else if (Target.starts_with("amd_gpu_"))
+          Triples.insert(C.getInputArgs().MakeArgString("amdgcn-amd-amdhsa"));
+      }
+      else
+        Triples.insert(C.getInputArgs().MakeArgString(Target));
```

Comment:
```suggestion
        if (Target.starts_with("intel_gpu_")) {
        Triples.insert(
            C.getInputArgs().MakeArgString("spir64_gen-unknown-unknown"));
          continue;  
        } else if (Target.starts_with("nvidia_gpu_")) {
          Triples.insert(C.getInputArgs().MakeArgString("nvptx64-nvidia-cuda"));
          continue;
        } else if (Target.starts_with("amd_gpu_")) {
          Triples.insert(C.getInputArgs().MakeArgString("amdgcn-amd-amdhsa"));
          continue;
        }
      }
      Triples.insert(C.getInputArgs().MakeArgString(Target));
```
Maybe something like this - or add an additional `else` for the `IsSYCL` condition.